### ISSUE TITLE
Add support for desktop actions

### ISSUE THREAD

**Issue by gsantner** [gmc-holle/xfdashboard#158]
Currently right clicking on  applications (both in application list/search, and on side-dock) does not reveal `desktop actions` (defined in `.desktop` files `[Desktop Action X]` sections). Would be nice if could show the actions too :).

The parser / Garcon does support this since June: <https://bugzilla.xfce.org/show_bug.cgi?id=11268#c9>

**Comment by gmc-holle** [comment]
It is already on my personal todo list ;) But as far as I know this support was added in 0.6.0 at garcon but Xfce 4.12 relies on garcon 0.2 as minimal version. So I want to support also the older versions of garcon. One way would be to add support for desktop action in xfdashboard only if garcon 0.6+ is installed at compile time or to add private utility functions which are used if older garcon version are installed at compile time. I prefer the latter so it takes more time to implement. And yes, currently I have less spare time to code :(

**Comment by gsantner** [comment]
I see you started on this already, nice! :D
Tell me when you need some feedback

**Comment by gmc-holle** [comment]
Feedback is welcome as it is now implemented but only for the applications shown in the quicklaunch.

**Comment by gsantner** [comment]
the dock?

**Comment by gmc-holle** [comment]
Yes, the dock at the left side.

**Comment by gsantner** [comment]
was there not the desktop wallpaper displayed (blurred) in the background in older versions? 
![bildschirmfoto_2017-10-23_13-09-24](https://user-images.githubusercontent.com/6735650/31886174-848ce6dc-b7f3-11e7-9b6c-04e89dbb5505.png)

**Comment by gsantner** [comment]
* Menu does not get closed when exiting overview (pressing windows key, pressing again opens xfdashboard with menu still open)
* Discuss?: Exit overview after selecting an action. Got e.g. "Open Downloads folder" as action on thunar, clicking on it does show an notification, but does not exit xfdashboard. Gnome shell does close when doing an action click

**Comment by gmc-holle** [comment]
Your comment about desktop wallpaper looks like a duplicate of #148

**Comment by gmc-holle** [comment]
> Menu does not get closed when exiting overview (pressing windows key, pressing again opens xfdashboard with menu still open)

This is a general bug of pop-up menus and not special to desktop action. Nevertheless it must be fixed.

> Discuss?: Exit overview after selecting an action. Got e.g. "Open Downloads folder" as action on thunar, clicking on it does show an notification, but does not exit xfdashboard. Gnome shell does close when doing an action click

Users may expect this to happen at xfdashboard too if they have seen it at Gnome Shell, so it should be mimiced.

**Comment by gmc-holle** [comment]
With commit bc6cee87160f546ea8281d9f342414c61753d599 it should quit (or suspend) when an action was selected at pop-up menu.

**Comment by gmc-holle** [comment]
Garcon provides an accessor function for "Actions" sind version 0.6.0. But it uses a hash-table to store the action with their action name as key. When returning the list of actions a list of hash-table keys is returned but their order will not represent the order as in the desktop entry file.

So I implemented a fallback code-path to retrieve the data about actions by reading the desktop entry file directly via GKeyFile when needed (and only once). The good of this solution is it is also usable for systems using Garcon prior to version 0.6.0 which do not have any accessor function for "Actions". And this code is reusable to implement keywords - see #159 - which no official release of Garcon supports so far :)

Except for the general pop-up bug - "not closing when suspending application" - this issue is fixed for me. Waiting for your feedback to close this issue finally.

**Comment by gsantner** [comment]
Ah nice :D. Basically the  Actions= linr gives you order and the keys :D

**Comment by gmc-holle** [comment]
Yes, indeed it does ;) It's time to file a bug for garcon and provide a patch to replace the hash-table with a GList for actions.

**Comment by gmc-holle** [comment]
Finally the general pop-up bug should be fixed now. Can you confirm? I would like to close this issue if it is ok :)

**Comment by gsantner** [comment]
Yes, indeed it does ;) It's time to file a bug for garcon and provide a patch to replace the hash-table with a GList for actions.


> I would, but I dont have the time to read into glib/gtk and the project as a whole. Keywords support was just Categories-Light 😉

**Comment by gmc-holle** [comment]
I was talking to me when I said it's time to file a bug ;) It's done: https://bugzilla.xfce.org/show_bug.cgi?id=13951

If desktop actions at xfdashboard are working for you and the pop-up menu bugs are considered as fixed, please tell me as I would be happy to close a issue. 😉

**Comment by gsantner** [comment]
Can confirm popup is fixed, and otherwise works in quicklaunch.

What is not impletmented is popup for search result and browsing results. Should there be another issue for that?

**Comment by gmc-holle** [comment]
Yes, please. :)

Am Donnerstag, den 26.10.2017, 03:09 -0700 schrieb Gregor Santner:
